### PR TITLE
[background] added events to ensure right page capture

### PR DESCRIPTION
**Pull request by gastonche**
PR pertaining to #15 .
I have added event listeners to all tab watching events that ensure if the tab url changes , it will reject the promise such that the watching action can't continue

### CODE REVIEW

File: src/util/tab-events.js [hunk 1]
```diff
@@ -1,5 +1,20 @@
 import eventToPromise from './event-to-promise'
 
+function tabChanged(tabId) {
+    return new Promise((resolve, reject) => {
+        browser.webNavigation.onCommitted.addListener((details)=>{
+            console.log(details);
```
Comment (obsidianart) [inline src/util/tab-events.js]: any reason to leave a console.log here?

File: src/util/tab-events.js [hunk 2]
```diff
@@ -1,5 +1,20 @@
...
+            if(details.tabId == tabId && details.frameId == 0)
+                reject();
+        });
+        browser.webNavigation.onHistoryStateUpdated.addListener((details)=>{
+            console.log(details);
```
Comment (obsidianart) [inline src/util/tab-events.js]: any reason to leave a console.log here?

File: src/util/tab-events.js [hunk 3]
```diff
@@ -1,5 +1,20 @@
 import eventToPromise from './event-to-promise'
 
+function tabChanged(tabId) {
+    return new Promise((resolve, reject) => {
+        browser.webNavigation.onCommitted.addListener((details)=>{
+            console.log(details);
+            if(details.tabId == tabId && details.frameId == 0)
+                reject();
```
Comment (obsidianart) [inline src/util/tab-events.js]: this might cause a double rejection since line 13 does the same thing. Even if this is allowed by the A+ pattern (the second rejection would be ignored) I wonder if there's any point in having a listener working but the promise rejected this way.

File: src/util/tab-events.js [hunk 4]
```diff
@@ -23,12 +38,15 @@ export function whenPageDOMLoaded({tabId}) {
...
-
-    // Workaround: we run a script at 'document_end' (= 'interactive')
-    return browser.tabs.executeScript(tabId, {
-        code: 'undefined', // Bogus code, anything works.
-        runAt: 'document_end',
-    })
+    return new Promise((resolve, reject) => {
+        // Workaround: we run a script at 'document_end' (= 'interactive')
+        browser.tabs.executeScript(tabId, {
+            code: 'undefined', // Bogus code, anything works.
```
Comment (obsidianart) [inline src/util/tab-events.js]: if you say```code: 'boguse_code_anything_works'``` you don't need the comment

File: src/util/tab-events.js [hunk 5]
```diff
@@ -23,12 +38,15 @@ export function whenPageDOMLoaded({tabId}) {
...
+            runAt: 'document_end',
+        }).then(resolve());
```
Comment (obsidianart) [inline src/util/tab-events.js]: this is incorrect and would call too early since you are invoking at declaration time, this should be
```
then(resolve) //if you want to pass the value of the then
```

or

```
.then(()=>resolve()) //if you want to resolve with empty value
```

Comment (obsidianart) [reply src/util/tab-events.js]: Also, there's no catch and this is discourage (and will be disallowed in node soon)

File: src/util/tab-events.js [hunk 6]
```diff
@@ -23,12 +38,15 @@ export function whenPageDOMLoaded({tabId}) {
...
+        
+        tabChanged(tabId).catch(() => reject());
```
Comment (obsidianart) [inline src/util/tab-events.js]: why is this needed at all?

Comment (gastonche) [reply src/util/tab-events.js]: This is needed, in order to check if the tab changes, so it rejects

File: src/util/tab-events.js [hunk 7]
```diff
@@ -40,18 +58,24 @@ export function whenPageLoadComplete({tabId}) {
...
-                    (changedTabId === tabId && status === 'complete'),
-            },
-            reject: {
-                event: browser.tabs.onRemoved,
-                filter: (closedTabId) => (closedTabId === tabId),
-                reason: {message: "Tab was closed before loading was complete."},
-            },
-        })
+        return new Promise((resolve, reject) => {
+            eventToPromise({
```
Comment (obsidianart) [inline src/util/tab-events.js]: @Treora I'll leave you evaluate if this library is correct or not here

File: src/util/tab-events.js [hunk 8]
```diff
@@ -40,18 +58,24 @@ export function whenPageLoadComplete({tabId}) {
...
+                    filter: (changedTabId, {status}) =>
+                        (changedTabId === tabId && status === 'complete'),
+                },
+                reject: {
+                    event: browser.tabs.onRemoved,
+                    filter: (closedTabId) => (closedTabId === tabId),
+                    reason: {message: "Tab was closed before loading was complete."},
+                },
+            }).then(()=>resolve())
+            .catch(()=>reject());
```
Comment (obsidianart) [inline src/util/tab-events.js]: this should probably be ```.catch(reject)``` otherwise the reject reason will be lost making quite hard to debug

File: src/util/tab-events.js [hunk 9]
```diff
@@ -40,18 +58,24 @@ export function whenPageLoadComplete({tabId}) {
...
+
+        tabChanged(tabId).catch(() => reject());
```
Comment (obsidianart) [inline src/util/tab-events.js]: why is this needed at all?

File: src/util/tab-events.js (hunk 8 above)
Comment (gastonche) [reply src/util/tab-events.js]: okay

File: src/util/tab-events.js (hunk 6 above)
Comment (obsidianart) [reply src/util/tab-events.js]: ok, interesting pattern, I'll leave it to @Treora  :)

**Comment by gastonche** [comment]
I will fix the corrections pointed out in the reviews, and update the PR,
I added the tabChanged method which returns a promise that rejects if the tab is changed, that way, the function waiting for a resolve will stop. In that way we  can avoid taking the snapshot of the wrong page
@obsidianart

**Comment by obsidianart** [comment]
@gastonche The approach is clear, but I've never seen promises used for an event based pattern so I'd leave Treora decide

File: src/util/tab-events.js (hunk 3 above)
Comment (gastonche) [reply src/util/tab-events.js]: My view on this was that, i have to listen to changes in the tabs url coming both from normal redirects like links clicked, and from changes triggered by the HTML 5 History API,  and reject in both cases, but i doubt there is ever a case  where those two can resolve at once and so this can only call reject once
@obsidianart

File: src/util/tab-events.js (hunk 5 above)
Comment (gastonche) [reply src/util/tab-events.js]: Terrible error

**Comment by Treora** [comment]
@obsidianart: It may not be very common, but promises are quite a fitting abstraction when waiting for an event to happen exactly once; especially when there is the need to stop waiting when another type of event happens. I wrote this small `eventToPromise` helper exactly because I ended up using this pattern several times.

@gastonche: looks like you got a technically sensible solution, roughly what I had in mind. I have a question the choice of events to listen to, and I think we can refactor the code to make it more understandable to a reader. Will comment inline.

**Comment by gastonche** [comment]
@Treora It would be nice to have it clearer for sure. What do you have in mind

File: src/util/tab-events.js [hunk 10]
```diff
@@ -1,57 +1,59 @@
...
+        browser.webNavigation.onCommitted.addListener((details)=>{
+            if(details.tabId == tabId && details.frameId == 0)
+                reject();
+        });
+        browser.webNavigation.onHistoryStateUpdated.addListener((details)=>{
+            if(details.tabId == tabId && details.frameId == 0)
+                reject();
+        });
+    });
+}
```
Comment (Treora) [inline src/util/tab-events.js]: My question about the choice of events: why do you choose to listen to onCommited and onHistoryStateUpdated? Is it ensured that these will capture all cases where a tab 'unloads' its page? There's a few more options in the [`webNavigation` API](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/webNavigation), and also listening to `tabs.onChange` to check for URL changes would be an option that comes to mind. Did you try or research any of these options?

I guess you may have chosen these because the activity logger also listens to them (when #50 is resolved). Listening to the same events could then indeed solve the problem of accidentally capturing the wrong page, so that's a smart solution. If it is possible it would be nice however to ensure we have a watertight solution, that could also be trusted to work for potential other scenarios.

File: src/util/tab-events.js [hunk 11]
```diff
@@ -1,57 +1,59 @@
 import eventToPromise from './event-to-promise'
 
+function tabChanged(tabId) {
+    return new Promise((resolve, reject) => {
+        browser.webNavigation.onCommitted.addListener((details)=>{
+            if(details.tabId == tabId && details.frameId == 0)
+                reject();
```
Comment (Treora) [inline src/util/tab-events.js]: Instead of rejecting, it would be more logical to resolve here, since the name is tabChanged (I'd call it `whenTabChanged` for consistency, or an even more accurate description). Where it is used code would then switch from `catch` to `then`.

File: src/util/tab-events.js (hunk 10 above)
Comment (Treora) [reply src/util/tab-events.js]: I think I made eventToPromise such that it accepts multiple event descriptions to listen for; instead of creating a new Promise each time, could the events defined in `tabChanged` be described in an array that can be added to the `reject` argument of eventToPromise instead?

I am thinking of something like this:
```
const tabChangedEvents = [
    {
        event: browser.tabs.onCommit
        filter: details => (details.tabId == tabId && details.frameId == 0)
    },
    .... same for onHistoryStateUpdated or other events ...
]
```

File: src/util/tab-events.js [hunk 12]
```diff
@@ -1,57 +1,59 @@
...
-                event: browser.tabs.onUpdated,
-                filter: (changedTabId, {status}) =>
-                    (changedTabId === tabId && status === 'complete'),
-            },
-            reject: {
-                event: browser.tabs.onRemoved,
-                filter: (closedTabId) => (closedTabId === tabId),
-                reason: {message: "Tab was closed before loading was complete."},
-            },
-        })
```
Comment (Treora) [inline src/util/tab-events.js]: So here the value of `reject` would be an array:
```
reject: [{
    event: browser.tabs.onRemoved,
    filter: ...,
    reason: ...,
    },
   ...tabChangedEvents,
]
```

File: src/util/tab-events.js (hunk 11 above)
Comment (Treora) [reply src/util/tab-events.js]: (forget about this comment, if we replace this `tabChanged` Promise with a list)

**Comment by gastonche** [comment]
Okay @Treora , i'll try to refactor the Code in order to use the event to promise approach

**Comment by gastonche** [comment]
@Treora I have refactored the code to use the eventToPromise Approach. But then I left the new Promise in the whenPageDOMLoaded method because the browser.tabs.executeScript could not be added to the list of tabsChangedEvents since it dows not have  an addlistener property and so will throw an error. I believe it's more readable this way.

**Comment by obsidianart** [comment]
@Treora I think event to Promise is a way to force a pattern within another, especially since you are waiting for rejection and not fullfilment. A promise rejection can come from many things (including syntax error).
Comparing the following I find this approach quite singular, also considering Node has events emitter (https://nodejs.org/api/events.html).
It might be my limit, but I always saw promises as a way to ask for something which you expect to happen, not to throw for something which you don't expect to happen.

```
//Promise
const waitForSmt = smt() 
waitForSmt.catch(doAction)
```

```
//Event
const waitFor = smt()
waitFor.on('smt', doAction)
```

```
//Callback
const waitForSmt = smt(doAction)
```

File: src/util/event-to-promise.js [hunk 13]
```diff
@@ -16,10 +16,10 @@ export default function eventToPromise({
     reject: rejectOpts,
 }) {
     // We can get a single options object, or an array of them, or undefined.
-    if (typeof resolveOpts !== 'Array') {
+    if (!Array.isArray(resolveOpts)) {
```
Comment (Treora) [inline src/util/event-to-promise.js:19]: Thanks for fixing, my bad! I wrote this code too hastily.

File: src/util/tab-events.js [hunk 14]
```diff
@@ -1,56 +1,55 @@
...
 // Resolve if or when the page is completely loaded.
 // Rejects if it is closed before that.
-// TODO reject if loading of the current page is aborted. (#15)
 export function whenPageLoadComplete({tabId}) {
     return browser.tabs.get(tabId).then(tab => {
         if (tab.status === 'complete')
             return // Resolve directly
 
-        return eventToPromise({
+    return eventToPromise({
```
Comment (Treora) [inline src/util/tab-events.js]: indent

File: src/util/tab-events.js [hunk 15]
```diff
@@ -65,17 +64,12 @@ export function whenTabActive({tabId}) {
     ).then(isActive => {
         if (isActive)
             return // Resolve directly
-
         return eventToPromise({
```
Comment (Treora) [inline src/util/tab-events.js]: lost a line here

File: src/util/tab-events.js [hunk 16]
```diff
@@ -1,56 +1,55 @@
...
+    },
+    {
+        event: browser.webNavigation.onHistoryStateUpdated,
+        filter: (details)=>(details.tabId == tabId && details.frameId == 0),
+        reason: "Tab URL changed before loading was complete"
+    },
+    {
+        event: browser.tabs.onRemoved,
+        filter: (closedTabId) => (closedTabId === tabId),
+        reason: {message: "Tab was closed before loading was complete."},
```
Comment (Treora) [inline src/util/tab-events.js]: These messages were different for each function, "before loading was complete" is a confusingly wrong message when printed by the whenTabActivated function. We can make them generic enough to match every of the functions, e.g. ".... before event occurred."
Also, it should be an object with a message field, not a string like you used above.

File: src/util/tab-events.js [hunk 17]
```diff
@@ -1,56 +1,55 @@
...
-    //             event: browser.webNavigation.onDOMContentLoaded,
-    //             filter: ({tabId: eventTabId}) => (eventTabId===tabId),
-    //         },
-    //         reject: {
-    //             event: browser.tabs.onRemoved,
-    //             filter: (closedTabId) => (closedTabId === tabId),
-    //             reason: {message: "Tab was closed before DOM was loaded."},
-    //         },
-    //     })
-    // })
```
Comment (Treora) [inline src/util/tab-events.js:25]: Removing the bulk of this commented code is fine, but I would still add an explanation why we use this `executeScript` approach: there is no tab.status==='interactive'.

**Review by Treora — COMMENTED** [review]
Looks pretty, at least within the given approach of using eventToPromise, which @obsidianart made me doubt about (let's first merge this before reconsidering the approach though).

Comments inline. And another small thing: forgot to remove the TODO at `whenTabActive`. Also we could perhaps rename it now to `whenPageActive`, since it is not about the tab in general anymore (we could have an argument determine whether to wait for the specific page or the tab in general, but let's do that another time, if we will need it).

**Comment by Treora** [comment]
> I think event to Promise is a way to force a pattern within another, especially since you are waiting for rejection and not fullfilment

@obsidianart: We are waiting for fulfilment. E.g. `whenPageLoaded` resolves when the page completed loading. It rejects if the page was closed before that ever happened.

I do understand and agree with your point that exceptions/rejections should be used for error handling, rather than for handling events that could be expected to happen. I am not sure if a page closing before it being loaded can reasonably be regarded as an exception.. For the user of the promise however, it probably does not matter if the promise rejected because of some really unforseeable error or because the page closed; either way what they waited for is not happening anymore.

I will think whether some event system can be more appropriate here, while still retaining the simplicity in usage, but I feel this promise approach is still acceptable.

**Comment by obsidianart** [comment]
Feel free to merge, I have no truth.

**Comment by Treora** [comment]
Great, looks ready to merge.
@gastonche: would you be okay with waiving your copyrights on your contributions and publish it in the public domain?

**Comment by gastonche** [comment]
no problem @Treora . I am more interested in seeing the fixes in the repo
